### DISCUSSION TITLE
SteppableRegionの受け取り方について

### DISCUSSION THREAD

**Discussion by Naoki-Hiraoka** [YutaKojio/hrpsys-base#45] [Q&A] [answered]
僕のソースコードの理解のどこかに間違えがあると思うのですが、指摘していただけますと幸いです。

1. l_rフラグには、現在の支持脚が入る(`right: 0, left: 1`). regionは、GaitGeneratorのworld座標系で表現される
2. 届いたdataのl_rフラグと、現在の支持脚が一致したら受け取る
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/AutoBalancer.cpp#L806-L809
3. lr_regionとsteppable_regionの、現在の遊脚のインデックスに格納する
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.h#L1812-L1827
4. GaitGeneratorの処理中、cur_legに、現在の遊脚が入る
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L807
5. lr_legionの、現在の遊脚のインデックスを消す **(3でせっかく格納したのに?)**
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L808
6. lr_regionの現在の支持脚のインデックスがtrueなら
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L809
steppable_regionの、現在の支持脚のインデックスに現在の遊脚用のstride_limitationを施す
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L857
7. steppable_regionの現在の支持脚のインデックスを用いて、現在の遊脚の着地位置をリミットする
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L1529

**Comment by YutaKojio** [comment, marked as answer]
自分で今見返してもなにやってるかすぐわからない処理になってました．．．

1. regionがどこの部分指しているかわからないけど，ROSから送るときは支持足相対で送って，hrpsys内では一歩遅れて一歩前の支持足相対（対象の遊脚の離床直前の足位置相対）で保持しています．
2. yes
3. yes. ただし，現在の遊脚と同じことだけど，6., 7. で使う場面を考えて next_support という名前にしています．
4. yes. 補足で，footstepのindexが変わる瞬間（両足支持期の真ん中）で行われます．
5. 4.でindexが変わっているので，対象にしたい足と逆の足のフラグを消していることになります．
6. yes
7. yes

タイミングを以下の手書きで示してみました．

![Screenshot from 2022-02-22 14-09-38](https://user-images.githubusercontent.com/11713954/155066974-ff1a2a30-f135-48fd-8295-89da7271be4a.png)

**Reply by Naoki-Hiraoka** [reply]
ありがとうございます。
- ROSから送るときは支持足相対で送って，hrpsys内では一歩遅れて一歩前の支持足相対（対象の遊脚の離床直前の足位置相対）で保持している
- footstepのindexが変わる瞬間（両足支持期の真ん中）で行われる

の2点を見落としていました。

具体的にllegの着地位置のsteppable regionを対象に考えると、

1. llegが支持脚のときに、次のllegの一歩のsteppable regionを`m_steppableRegion`として受け取る。l_rフラグには、1が入る(`right: 0, left: 1`).`m_steppableRegion.data.region`は、現在のlleg相対の座標系で表現される
2. 届いたdataのl_rフラグと、現在の支持脚が一致したら受け取る
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/AutoBalancer.cpp#L806-L809
3. lr_regionとsteppable_regionの、rlegのインデックスに格納する(next_support という名前にしている)
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.h#L1812-L1827
4. footstepのindexが変わる瞬間（両足支持期の真ん中）、4,5,6の処理が行われる.
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L806 llegが遊脚になるので、cur_legにllegが入る
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L807
5. (削除)
6. lr_regionのrlegのインデックスがtrueなら
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L809
steppable_regionの、rlegのインデックスにllegのstride_limitationを施す
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L857
7. 以降、steppable_regionのrlegのインデックスを用いて、llegの着地位置をリミットする
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L1529
8. footstepのindexが変わる瞬間（両足支持期の真ん中）、llegが支持脚になり、lr_legionのrlegのインデックスを消す
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L808
